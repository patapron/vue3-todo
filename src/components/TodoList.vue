<template>
  <v-card class="mx-auto" width="300">
    <v-text-field
      label="New todo"
      hide-details
      v-model="newTodoName"
      v-on:keyup.enter="addNewTodo"
    ></v-text-field>
    <v-list>
      <!-- active -->
      <ListGroup
        v-if="name1"
        :list="getCompleted(false)"
        :name="props.name1"
        @change-event="changeTodo"
      ></ListGroup>
      <!-- completed -->
      <ListGroup
        v-if="name2"
        :list="getCompleted(true)"
        :name="props.name2"
        @change-event="changeTodo"
      ></ListGroup>
    </v-list>
  </v-card>
</template>

<script>
import { useStore } from "vuex";
import { computed, ref, defineComponent, provide } from "vue";

import ListGroup from "../components/ListGroup.vue";
import mitt from "mitt";

export default defineComponent({
  name: "TodoList",
  components: {
    ListGroup,
  },
  props: {
    filterValue: {
      type: Boolean,
      default: false,
    },
    name1: String,
    name2: String,
  },
  setup(props) {
    const store = useStore();
    const all = computed(() => store.state.todos);
    const getCompleted = computed(() => store.getters.getCompleted);
    const newTodoName = ref("");
    const bus = mitt();
    bus.on("delete-event", (evt) => {
      store.dispatch("deleteTodo", evt);
    });
    bus.on("change-event", (evt) => {
      store.dispatch("changeTodo", evt);
    });
    provide("bus", bus);

    function addNewTodo() {
      let newTodo = { id: null, name: newTodoName.value, completed: false };
      store.dispatch("addTodo", newTodo);
      newTodoName.value = "";
    }

    return {
      all,
      newTodoName,
      addNewTodo,
      getCompleted,
      props,
    };
  },
});
</script>
