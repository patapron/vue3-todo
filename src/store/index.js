import { createStore } from "vuex";

export default createStore({
  state: {
    todos: [],
    idCounter: 0,
  },
  mutations: {
    setTodos(state, payload) {
      state.todos = payload;
    },
    counterIncrement(state) {
      state.idCounter++;
    },
  },
  actions: {
    getTodos: (state) => {
      // fetch(`https://todos.com/api/getTodos`)
      //   .then((res) => {
      //     return res.json();
      //   })
      //   .then((response) => {
      //     state.commit("setTodos", response);
      //   });
      const response = [];
      state.commit("setTodos", response);
    },
    addTodo({ commit, state }, todo) {
      let tempTodo = [...state.todos];
      todo.id = state.idCounter;
      commit("counterIncrement");
      tempTodo.push(todo);
      commit("setTodos", tempTodo);
    },

    changeTodo({ commit, state }, todo) {
      let tempTodos = [...state.todos];
      tempTodos.forEach(element => {
        if(element.id === todo.id){
          element.name = todo.name
          element.completed = todo.completed
          return
        }
      });
      commit("setTodos", tempTodos);
    },

    deleteTodo({ commit, state }, todo) {
      let tempTodos = state.todos.filter((element) => {
        return element.id !== todo.id;
      });
      commit("setTodos", tempTodos);
    },

    deleteCompleted({ commit, state }) {
      let tempTodos = state.todos.filter((element) => {
        return !element.completed;
      });
      commit("setTodos", tempTodos);
    },
  },
  getters: {
    getCompleted: (state) => (value) => {
      let result = [];
      if (state.todos.length) {
        result = state.todos.filter((element) => element.completed === value);
      }
      return result;
    },
  },
});
