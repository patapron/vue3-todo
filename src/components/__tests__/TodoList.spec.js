/* eslint-disable prettier/prettier */
import TodoList from "@/components/TodoList.vue";
import Vuex from "vuex";
import { shallowMount } from "@vue/test-utils";
import storeConfig from "../../store/index";
import { mutations } from "@/store/index.js";
import { cloneDeep } from "lodash";
import { createLocalVue } from "@vue/test-utils";

import store from "@/store/index.js";

describe("TodoList", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(TodoList);
  });

  it("renders correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('"counterIncrement" increments "state.idCounter" by 1', () => {
    expect(store.state.idCounter).toBe(0);
    store.commit("counterIncrement");
    expect(store.state.idCounter).toBe(1);
  });

  // it('"setTodos" set state.todos with whatever', () => {
  //   expect(store.state.todos).toEqual([]);
  //   store.commit("setTodos", [{ id: 0, name: "test", completed: false }]);
  //   expect(store.state.todos).toEqual([
  //     { id: 0, name: "test", completed: false },
  //   ]);
  // });

  it("addTodo", async () => {
    store.dispatch("addTodo", { id: null, name: "test", completed: false });
    expect(store.state.todos).toEqual([
      { id: 1, name: "test", completed: false },
    ]);
  });

  it("deleteTodo", async () => {
    store.dispatch("deleteTodo", { id: 1, name: "test", completed: false });
    expect(store.state.todos.length).toBe(0);
  });
});

it("changeTodo", async () => {
  store.dispatch("addTodo", { id: null, name: "test", completed: false });
  store.dispatch("changeTodo", { id: 2, name: "test2", completed: true });

  expect(store.state.todos[0]).toEqual({
    id: 2,
    name: "test2",
    completed: true,
  });
});

// import { createStore } from 'vuex'

// describe("TodoList", () => {

//   const store = createStore({
//     state() {
//       return {
//         idCounter: 0
//       }
//     },
//     mutations: {
//       counterIncrement(state) {
//         state.idCounter += 1
//       }
//     }
//   })
//   beforeEach(() => {
//   });

//   it("vuex", async () => {
//     const wrapper = mount(TodoList, {
//       global: {
//         plugins: [store],
//       },
//     });
//     console.log("STORE", store)
//     expect(store.state.idCounter).toBe(0);
//     store.commit("counterIncrement");
//     expect(store.state.idCounter).toBe(1);
//   });
// });
