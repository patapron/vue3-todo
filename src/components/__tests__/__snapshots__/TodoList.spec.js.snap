// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders correctly 1`] = `
<v-card class="mx-auto" width="300">
  <v-text-field label="New todo" hide-details="" modelvalue=""></v-text-field>
  <v-list>
    <!-- active -->
    <!--v-if-->
    <!-- completed -->
    <!--v-if-->
  </v-list>
</v-card>
`;

exports[`TodoList renders correctly 1`] = `
<v-card class="mx-auto" width="300">
  <v-text-field label="New todo" hide-details="" modelvalue=""></v-text-field>
  <v-list>
    <!-- active -->
    <!--v-if-->
    <!-- completed -->
    <!--v-if-->
  </v-list>
</v-card>
`;
